lib/pure/strinterp.nim

```diff
@@ -0,0 +1,320 @@
+  if fmtp == "":
+    return
+  var t = ""
+  var minus = 1
+  var idx = 0
+  idx += fmtp.parseWhile(t, {'-'}, idx)
+  if t == "-":
+    minus = -1
+  idx += fmtp.parseWhile(t, {'0'}, idx)
+  if t == "0":
+    result.fillChar = '0'
+  idx += fmtp.parseWhile(t, {'0'..'9'}, idx)
+  if t != "":
+    result.maxLen = minus * parseInt(t)
+  idx += fmtp.skipWhile({'.'}, idx)
+  idx += fmtp.parseWhile(t, {'0'..'9'}, idx)
+  if t != "":
+    result.prec = parseInt(t)
+
+proc handleIntFormat(exp: string, fmtp: string, radix: int, lowerCase = false): NimNode {.compileTime.} =
+  let (maxLen, fillChar) = parseIntFmt(fmtp)
+  result = newCall(bindSym"intToStr", parseExpr(exp), newLit(radix), newLit(maxLen), newLit(fillChar), newLit(lowerCase))
+
+proc handleDFormat(exp: string, fmtp: string): NimNode {.compileTime.} =
+  result = handleIntFormat(exp, fmtp, 10)
+
+proc handleXFormat(exp: string, fmtp: string, lowerCase: bool): NimNode {.compileTime.} =
+  result = handleIntFormat(exp, fmtp, 16, lowerCase)
+
+proc handleFFormat(exp: string, fmtp: string): NimNode {.compileTime.} =
+  var (maxLen, prec, fillChar) = parseFloatFmt(fmtp)
+  result = newCall(bindSym"floatToStr", parseExpr(exp), newLit(maxLen), newLit(prec), newLit('.'), newLit(fillChar), newLit(false))
+
+proc handleEFormat(exp: string, fmtp: string): NimNode {.compileTime.} =
+  var (maxLen, prec, fillChar) = parseFloatFmt(fmtp)
+  result = newCall(bindSym"floatToStr", parseExpr(exp), newLit(maxLen), newLit(prec), newLit('.'), newLit(fillChar), newLit(true))
+
+proc handleSFormat(exp: string, fmtp: string): NimNode {.compileTime.} =
+  var (maxLen, _) = parseIntFmt(fmtp)
+  if maxLen == 0:
+    result = parseExpr("$(" & exp & ")")
+  else:
+    result = newCall(bindSym"alignStr", parseExpr("$(" & exp & ")"), newLit(maxLen), newLit(' '))
+
+proc handleFormat(exp: string, fmt: string, nodes: var seq[NimNode]) {.compileTime} =
+  if fmt[1] == '%':
+    nodes.add(parseExpr("$(" & exp & ")"))
+    nodes.add(newLit(fmt[1..^1]))
+  else:
+    const formats = {'d', 'f', 's', 'x', 'X', 'e'}
+    var idx = 1
+    var fmtm = ""
+    var fmtp = ""
+    while idx < fmt.len:
+      if fmt[idx].isAlphaAscii:
+        if fmt[idx] in formats:
+          fmtm = $fmt[idx]
+          fmtp = fmt[1..idx-1]
+        break
+      inc idx
+    if fmtm == "":
+      nodes.add(parseExpr("$(" & exp & ")"))
+      nodes.add(newLit(fmt))
+    else:
+      case fmtm
+      of "d":
+        nodes.add(handleDFormat(exp, fmtp))
+      of "x", "X":
+        nodes.add(handleXFormat(exp, fmtp, lowerCase = fmtm == "x"))
+      of "f":
+        nodes.add(handleFFormat(exp, fmtp))
+      of "e":
+        nodes.add(handleEFormat(exp, fmtp))
+      of "s":
+        nodes.add(handleSFormat(exp, fmtp))
+      else:
+        quit "Unknown format \"" & fmtm & "\""
+      nodes.add(newLit(fmt[idx+1..^1]))
+
+macro fmt*(fmt: static[string]): untyped =
+  ## String interpolation macro with scala-like format specifiers.
+  ## Knows about:
```

Comment:
Note: An empty line is required for a newline to be present in the generated docs, currently ``Knows about`` will be on the same line as the above sentence.